zcash_primitives/src/transaction/txid.rs

```diff
@@ -0,0 +1,580 @@
+use std::borrow::Borrow;
+use std::convert::TryFrom;
+use std::io::Write;
+
+use blake2b_simd::Hash as Blake2bHash;
+use byteorder::{LittleEndian, WriteBytesExt};
+use ff::PrimeField;
+use group::GroupEncoding;
+use orchard::bundle::{self as orchard};
+
+use crate::consensus::{BlockHeight, BranchId};
+
+use super::{
+    blake2b_256::HashWriter,
+    components::{
+        amount::Amount,
+        orchard as ser_orch,
+        sapling::{self, OutputDescription, SpendDescription},
+        transparent::{self, TxIn, TxOut},
+    },
+    Authorization, Authorized, TransactionDigest, TransparentDigests, TxDigests, TxId, TxVersion,
+};
+
+#[cfg(feature = "zfuture")]
+use super::{
+    components::tze::{self, TzeIn, TzeOut},
+    TzeDigests,
+};
+
+/// TxId tree root personalization
+const ZCASH_TX_PERSONALIZATION_PREFIX: &[u8; 12] = b"ZcashTxHash_";
+
+// TxId level 1 node personalization
+const ZCASH_HEADERS_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdHeadersHash";
+const ZCASH_TRANSPARENT_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdTranspaHash";
+const ZCASH_SAPLING_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdSaplingHash";
+const ZCASH_ORCHARD_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdOrchardHash";
+#[cfg(feature = "zfuture")]
+const ZCASH_TZE_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdTZE____Hash";
+
+// TxId transparent level 2 node personalization
+const ZCASH_PREVOUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdPrevoutHash";
+const ZCASH_SEQUENCE_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdSequencHash";
+const ZCASH_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdOutputsHash";
```

Comment:
(as in https://zips.z.cash/zip-0244#t-3b-sapling-outputs-digest)

```suggestion
const ZCASH_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdSOutputHash";
```